libraries/tickfs/src/flash_controller.rs

```diff
@@ -0,0 +1,100 @@
+//! The Flash Controller interface with hardware
+
+use crate::error_codes::ErrorCode;
+
+/// Implementation required for the flash controller hardware. This
+/// should read, write and erase flash from the hardware using the
+/// flash controller.
+///
+/// This is the public trait for the Flash controller implementation.
+///
+/// The PageSize 'P' must be the smallest size that can be erased in
+/// a single operation.
+/// The start and end address of the FlashController must be aligned
+/// to the PageSize.
+/// All `region_number`s and `address`es are offset from zero. If you
+/// want to use flash that doesn't start at zero, or is a partition
+/// offset from the start of flash you will need to add that offset
+/// to the values in your implementation.
+///
+/// The boiler plate for an implementation will look something like this
+///
+/// ```rust
+/// use tickfs::error_codes::ErrorCode;
+/// use tickfs::flash_controller::FlashController;
+///
+/// #[derive(Default)]
+/// struct FlashCtrl {}
+///
+/// impl FlashCtrl {
+///     fn new() -> Self {
+///         Self { /* fields */ }
+///     }
+/// }
+///
+/// impl FlashController for FlashCtrl {
+///     fn read_region(&self, region_number: usize, offset: usize, buf: &mut [u8]) -> Result<(), ErrorCode> {
+///         unimplemented!()
+///     }
+///
+///     fn write_region(&self, address: usize, buf: &[u8]) -> Result<(), ErrorCode> {
+///         unimplemented!()
+///     }
+///
+///     fn erase_region(&self, region_number: usize) -> Result<(), ErrorCode> {
+///         unimplemented!()
+///     }
+/// }
+/// ```
+pub trait FlashController {
```

Comment:
Why do we need a new flash interface, rather than reusing the interface in `kernel/src/hil/flash.rs`? I understand that removing static lifetimes was one motivation, but this makes additional modifications (operating on regions instead of pages, offering offset support) -- should we just be making those changes to the HIL?

Reply:
The idea is that TickFS can be split out as a separate crate (like the Tock registers). The hope is then that others can use it as there doesn't seem to be anything else in Rust for storing data on flash. That is why a new interface is added.

It is probably worth improving the current Tock Flash HIL, as allowing address writes is pretty useful. I think that can be done separately though.